### DISCUSSION TITLE
One flow across multiple PCAPs

### DISCUSSION THREAD

**Discussion by JKyang09** [zeek/zeek#1716] [Ideas]
Currently, I have 500G of pcap data contains 30 minutes of network traffic.  Zeek should have a command to process the pcap folder, just like Suricata, and ensure that the flow across multiple PCAPs will not be split into multiple records

**Comment by 0xxon** [comment]
Thank you for your issue. Since this is a feature idea, I am going to move this to a discussion, where we currently put these.

**Comment by petiepooo** [comment]
@JKyang09 with a little scripting, you can output a pcap file global header, then concatenate all pcaps (minus each file's global header) into stdout, and feed it all to zeek to read the pcap from stdin.  Something like:

> (dd if=file1.pcap ibs=24 count=1; for pfile in *.pcap; do dd if=$pfile ibs=24 skip=1; done) | zeek -r -

As long as the shell glob orders the files correctly, the result should be that zeek sees one continuous pcap of 500GB.

**Reply by vpax** [reply]
You also might want to consider using `mergecap`, which will construct the aggregate pcap (in various formats) and merge each of the components by timestamp, which you might need if they have some temporal overlap.

**Reply by petiepooo** [reply]
Good point.  Mergecap allows the use of '-' as the output file to send to stdout.  This is a better solution than dd.

**Reply by petiepooo** [reply]
In fact, since this feature exists within mergecap, I am opposed to adding it to zeek.  Let pcap tools manage pcaps so zeek can stick to parsing packets.  Sorry, JKyang09...

**Reply by CyberTaoFlow** [reply]
A nice solution for this is available from CIRCL called pcapdj - https://github.com/CIRCL/pcapdj

**Comment by ckreibich** [comment]
There is definitely scope for a nice pcap batch-processing Zeek plugin. In most day-to-day settings `mergecap ... -w -| zeek -r -` will nail it, but mergecap requires all pcaps to be present at the time of invocation, and doesn't deal well with large numbers of files.

**Reply by petiepooo** [reply]
A few years ago I wrote a user-space process that would watch a pcap directory on a HDD (pcaps were written by netsniff-ng), and concatenate them all to a stdout stream in near real-time as they were written.  It would run for a day or more at about 700-800Mbps, IIRC, but depended on reading the file while its contents were still in the page-cache.  If the cache ever got flushed, or for some reason it fell far enough behind that the writes were flushed and the page was reclaimed by the time my process was ready to read, it could never catch up and I'd lose data.  It might fair better with an SSD as a backing store, but the HDD I had could not handle sustained concurrent reads and writes at line speed.

This was in snort2 days, and I had another user-space process to split that pcap iostream into multiple partial iostreams to send to snort processes much like pfring could do in kernel-space.  With af_packet and snort3 now, it's no longer needed and the project was abandoned.

That said, I can see the utility of a plugin like that for zeek.  If the intent is to watch a directory where another process is writing pcaps, please include a cursor of some sort to track the most recently read file/timestamp so you can restart zeek without losing or duplicating logs.

**Reply by ckreibich** [reply]
Great comments, thank you! I totally agree that as you get way into this you're asking for the wheels to fall off in all sorts of interesting ways. Also consider ramdisks etc — lots of ways to set these things up, depending on the use case.
